### DISCUSSION TITLE
Generate Ruby/JRuby/JRubyFX controller skeleton

### DISCUSSION THREAD

**Discussion by byteit101** [gluonhq/scenebuilder#595] [Ideas]
I've been the maintainer for JRubyFX (JavaFX bindings for JRuby), for just about a decade. I'd like to add Ruby/JRuby/JRubyFX to the controller skeleton generation now that we use the standard FXMLLoader since JRubyFX 2.0. I made a quick hack at https://github.com/byteit101/scenebuilder/tree/jruby that generates ok skeletons. Do note that due to Ruby's dynamism, we hack the constructors and Java-Ruby interface so that it looks like a normal Ruby object when building objects in the "JRubyFX way" (the `fxml "blah.fxml"` line). This means that the usage is slightly different than Java/Kotlin

**Comment by Oliver-Loeffler** [comment]
Hi @byteit101 ,
this is an awesome idea.  Would you mind in creating a PR for that? Depending on the state of work, some adjustments are required (e.g. adding a test, updating license header etc.). Would you support that approach? For me, technically this looks good so far and I'll try this out later.

**Reply by byteit101** [reply]
I can do that, though I think the current split of java/kotlin in the parent class isn't built well for rubyisms (java imports are usually just references INSIDE the class, variables defined automatically, `{}` vs `do end`, `//` vs `#`, etc), so I might just pull it further apart in the class hierarchy, like base -> {JRuby, CStyle -> {Java, Kotlin}}

I don't see a basic application/launcher skeleton, that isn't a thing, correct?

**Reply by Oliver-Loeffler** [reply]
There is no basic application / skeleton yet. But, if this makes sense for the way how JavaFX is used in conjunction with Ruby, we could add 2 options for Ruby. One could be the application skeleton, another one the controller skeleton.

I know, this approach is probably not perfect, but this would work for the beginning. Alternatively, we could add a checkbox so that for JRubyFX, one could check 'Application Skeleton'. This checkbox would remain disabled for Java and Kotlin. Instead of individual checkboxes, this could be a combo box with options or a scrollable list field. 

![AppSkeleton](https://user-images.githubusercontent.com/22102800/198106557-31763f92-a99c-4c0b-aa0e-885eaf0d577c.png)

**Reply by Oliver-Loeffler** [reply]
It would be favorable to not extend `AbstractSkeletonCreator`. For Kotlin and Java this was good way. For Ruby there could be another implementation. The important thing is, that the new SkeletonCreator consumes the `SkeletonContext`. Each SkeletonCreator can be replaced or described as a function, taking a SkeletonContext and turning it into a String. There is actually no class tree or lots of inheritance needed.

Something like that:

```java
Function<SkeletonContext, String> jRubyConverter= new SkeletonCreatorJRuby();
```

The existing creators can be updated to make use of the new interface and for JRuby or other languages one could fully rely on reusing the SkeletonContext.

A new interface:
```java
@FunctionalInterface
public interface SkeletonConverter {
    String createFrom(SkeletonContext context);
}
```

The existing creators would implement that interface via AbstractSkeletonCreator:
```java
abstract class AbstractSkeletonCreator implements SkeletonConverter {
}
```

The existing SkeletonCreator could be reworked like below.
With that, you could ignore the AbstractSkeletonCreator class and implement a suitable version straight forward for JRuby.

```java
class SkeletonCreator {
    String createFrom(SkeletonContext context) {
        return switch (context.getSettings().getLanguage()) {
            case JAVA -> new SkeletonCreatorJava().createFrom(context);
            case KOTLIN -> new SkeletonCreatorKotlin().createFrom(context);
            case JRUBY -> new SkeletonCreatorJRuby().createFrom(context);
        };
    }
}
```
What do you think?

**Reply by byteit101** [reply]
That could work, I'll play around this evening with that. For reference, here is the demo fxml sample for JRubyFX: https://github.com/jruby/jrubyfx/blob/master/samples/fxml/Demo.rb .  There are also some ports of java samples in contrib/

Application skeleton would be something like:

```rb
require 'jrubyfx'

# the controller skeleton goes here

fxml_root File.dirname(__FILE__) # or wherever you save the fxml file to

class FXMLExample < JRubyFX::Application
  def start(stage)
    ControllerName.load_into(stage)
    #stage.title = "FXML Application"
    stage.show
  end
end

FXMLExample.launch
```

**Reply by Oliver-Loeffler** [reply]
That would be awesome. Just thinking about adding similar functionality for Scala! JavaFX exists there as well.

**Reply by byteit101** [reply]
BTW, if you'd like to test this locally, you can either install jruby (rvm or installer), or just use the [complete jar](https://repo1.maven.org/maven2/org/jruby/jruby-complete/9.3.9.0/jruby-complete-9.3.9.0.jar) which has no prerequisites except a JVM:
```
export GEM_HOME=/tmp/gems
export GEM_PATH=/tmp/gems
java -jar jruby-complete.jar -S gem install jrubyfx

# later
java -jar jruby-complete.jar your_file.rb
```

**Reply by Oliver-Loeffler** [reply]
Thanks for sharing, will do so!

**Comment by Oliver-Loeffler** [comment]
Just for reference, this is how it looks, based on the complex application template from Scene Builder:

default view:

```ruby

class PleaseProvideControllerClassName
    include JRubyFX::Controller

    # fxml 'Untitled'

    # @centerPane: AnchorPane
    # @mainMenu: Menu
    # @mainMenuBar: MenuBar
    # @mainMenuItem: MenuItem
    # @rightPaneLabel: Label
    # @x1: Font
    # @x2: Color
    # @x3: Font
    # @x4: Color

    def mainAction(event) # event: ActionEvent

    end

end
```

with comments:
```ruby
#
# Sample Skeleton for 'Untitled' Controller Class
# 


class PleaseProvideControllerClassName
    include JRubyFX::Controller

    # fxml 'Untitled'

    # @centerPane: AnchorPane (Value injected by FXMLLoader & JRubyFX)
    # @mainMenu: Menu (Value injected by FXMLLoader & JRubyFX)
    # @mainMenuBar: MenuBar (Value injected by FXMLLoader & JRubyFX)
    # @mainMenuItem: MenuItem (Value injected by FXMLLoader & JRubyFX)
    # @rightPaneLabel: Label (Value injected by FXMLLoader & JRubyFX)
    # @x1: Font (Value injected by FXMLLoader & JRubyFX)
    # @x2: Color (Value injected by FXMLLoader & JRubyFX)
    # @x3: Font (Value injected by FXMLLoader & JRubyFX)
    # @x4: Color (Value injected by FXMLLoader & JRubyFX)

    def mainAction(event) # event: ActionEvent

    end

end
```

full, with comments:
```ruby
#
# Sample Skeleton for 'Untitled' Controller Class
# 


class PleaseProvideControllerClassName
    include JRubyFX::Controller

    fxml 'Untitled'

    # @centerPane: AnchorPane (Value injected by FXMLLoader & JRubyFX)
    # @mainMenu: Menu (Value injected by FXMLLoader & JRubyFX)
    # @mainMenuBar: MenuBar (Value injected by FXMLLoader & JRubyFX)
    # @mainMenuItem: MenuItem (Value injected by FXMLLoader & JRubyFX)
    # @rightPaneLabel: Label (Value injected by FXMLLoader & JRubyFX)
    # @x1: Font (Value injected by FXMLLoader & JRubyFX)
    # @x2: Color (Value injected by FXMLLoader & JRubyFX)
    # @x3: Font (Value injected by FXMLLoader & JRubyFX)
    # @x4: Color (Value injected by FXMLLoader & JRubyFX)

    def mainAction(event) # event: ActionEvent

    end

    # Called by JRubyFX after FXML loading is complete. Different from Java, same as normal Ruby
    def initialize()
        raise 'fx:id="centerPane" was not injected: check your FXML file "Untitled".' if @centerPane.nil?
        raise 'fx:id="mainMenu" was not injected: check your FXML file "Untitled".' if @mainMenu.nil?
        raise 'fx:id="mainMenuBar" was not injected: check your FXML file "Untitled".' if @mainMenuBar.nil?
        raise 'fx:id="mainMenuItem" was not injected: check your FXML file "Untitled".' if @mainMenuItem.nil?
        raise 'fx:id="rightPaneLabel" was not injected: check your FXML file "Untitled".' if @rightPaneLabel.nil?
        raise 'fx:id="x1" was not injected: check your FXML file "Untitled".' if @x1.nil?
        raise 'fx:id="x2" was not injected: check your FXML file "Untitled".' if @x2.nil?
        raise 'fx:id="x3" was not injected: check your FXML file "Untitled".' if @x3.nil?
        raise 'fx:id="x4" was not injected: check your FXML file "Untitled".' if @x4.nil?

    end

end
```

**Reply by byteit101** [reply]
> `fxml 'Untitled'`

Oh hmm, that should be an error/more obvious to fix

**Reply by Oliver-Loeffler** [reply]
Ah, thats not an error. Its just because there was no name defined for the FXML. Once the FXML is saved, The untitled is replaced.

It works, here the examples:
![1](https://user-images.githubusercontent.com/22102800/198113333-08342f59-cde3-4a4e-b6fa-36b0d72afd4c.PNG)
![3](https://user-images.githubusercontent.com/22102800/198113348-25547c8b-1c0b-44a1-b7cd-ca5d07a7c2e2.PNG)

**Comment by byteit101** [comment]
I have opened https://github.com/gluonhq/scenebuilder/pull/596 as a draft PR

**Comment by Oliver-Loeffler** [comment]
This is a good idea and effectively it's probably a nice thing to add similar functionality for ScalaFX.
@rladstaetter is using JavaFX in LogoRRR which is a Scala application. May be @rladstaetter can help here.

But first we need the draft PR #596 developed so that we can merge it.

**Reply by byteit101** [reply]
I think everything is good in the PR, so I unmarked it as draft. Let me know if there is something else I should do

**Reply by rladstaetter** [reply]
I'm using plain Scala for JavaFX interfacing, but there is something quite similar to JRubyFX named ScalaFX (https://github.com/scalafx/scalafx) maintained by @jpsacha. 

Looking at the pr, I think I could contribute code to generate plain scala code, although it would be heavily flavored by my approach to do things, which may or may not be the correct way of doing it ...

**Reply by Oliver-Loeffler** [reply]
Both ways are valid ways. Hence why not starting with the Scala without using ScalaFX. If that works out well, we could add a version which makes use of ScalaFX.

**Reply by rladstaetter** [reply]
I hacked something together here: https://github.com/rladstaetter/scenebuilder/tree/pr-skeleton-scala - it seems to work but surely needs some guidance for a pr / code review. Please advise how to proceed.

**Reply by Oliver-Loeffler** [reply]
I just skimmed over the code and it looks good for me. There are even tests, which is even better. I'll try this out tonight. The existing Ruby PR already anticipates that possibly other languages get added, hence there are few changes compared to mainline Scene Builder. Depending on which PR gets merged first, this might require some rework. One difference is, that there will be no need anymore to inherit from `AbstractSkeletonCreator`. This helps when the language is very similar to Java but if not it does not help. Instead an interface `SkeletonConverter` was added which has one method `public String createFrom(SkeletonContext context)`. So there an be a converter for any language, it starts all with the FXOMContext.

**Reply by byteit101** [reply]
Do let me know if there is something I should fix, as I think my PR is ready #596

**Reply by Oliver-Loeffler** [reply]
I will look into the code and start a review process in Github and point out if there are locations which require some rework.

**Comment by Oliver-Loeffler** [comment]
Great news @byteit101, the JRuby PR made it now into the mainline. Unfortunately, this means some rework for @rladstaetter as the structure changed a little. Not too much, but it wont merge automatically anymore.
